core/imds/src/main/java/software/amazon/awssdk/imds/internal/EndpointProvider.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.imds.internal;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.function.Supplier;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.SdkSystemSetting;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.profiles.Profile;
+import software.amazon.awssdk.profiles.ProfileFile;
+import software.amazon.awssdk.profiles.ProfileFileSystemSetting;
+import software.amazon.awssdk.profiles.ProfileProperty;
+
+/**
+ * Endpoint Provider Class which contains methods for endpoint resolution.
+ */
+@SdkInternalApi
+public class EndpointProvider {
+
+    private static final String EC2_METADATA_SERVICE_URL_IPV4 = "http://169.254.169.254";
+
+    /** Default IPv6 endpoint for the Amazon EC2 Instance Metadata Service. */
+    private static final String EC2_METADATA_SERVICE_URL_IPV6 = "http://[fd00:ec2::254]";
+
+    private final Supplier<ProfileFile> profileFile;
+    private final String profileName;
+
+    private EndpointProvider(Builder builder) {
+        this.profileFile = builder.profileFile;
+        this.profileName = builder.profileName;
+    }
+
+    public String getEndpoint(EndpointMode endpointMode) {
+
+        String endpointOverride = getEndpointOverride();
+        if (endpointOverride != null) {
+            return endpointOverride;
+        }
+
+        EndpointMode finalEndpointMode = resolveEndpointMode(endpointMode);
+        if (finalEndpointMode == EndpointMode.IPV4) {
+            return EC2_METADATA_SERVICE_URL_IPV4 ;
+        } else if (finalEndpointMode == EndpointMode.IPV6) {
+            return EC2_METADATA_SERVICE_URL_IPV6 ;
+        } else {
+            throw SdkClientException.create("Unknown endpoint mode: " + endpointMode);
+        }
+    }
+
+    public EndpointMode getEndpointMode() {
+
+        Optional<String> endpointMode = SdkSystemSetting.AWS_EC2_METADATA_SERVICE_ENDPOINT_MODE.getNonDefaultStringValue();
+        if (endpointMode.isPresent()) {
+            return EndpointMode.fromValue(endpointMode.get());
```

Comment:
Are these values case-sensitive in the SEP? It feels weird to accept iPv4, for example.

Reply:
I have provided an ENUM for the same, when user enters endpoint Mode in the builder. That uses IPV4 or IPV6 as standards. But the SdkSystemSettings and ConfigFile is where the existing functionality , converts to lower case always to check